x-pack/plugins/apm/server/lib/service_map/get_service_anomalies.ts

```diff
@@ -0,0 +1,158 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+import { intersection } from 'lodash';
+import { leftJoin } from '../../../common/utils/left_join';
+import { rangeFilter } from '../helpers/range_filter';
+import { Job as AnomalyDetectionJob } from '../../../../ml/server';
+import { PromiseReturnType } from '../../../typings/common';
+import { IEnvOptions } from './get_service_map';
+import { APM_ML_JOB_GROUP_NAME } from '../../../common/ml_job_constants';
+
+type ApmMlJobCategory = NonNullable<ReturnType<typeof getApmMlJobCategory>>;
+const getApmMlJobCategory = (
+  mlJob: AnomalyDetectionJob,
+  serviceNames: string[]
+) => {
+  const apmJobGroups = mlJob.groups.filter(
+    (groupName) => groupName !== APM_ML_JOB_GROUP_NAME
+  );
+  if (apmJobGroups.length === mlJob.groups.length) {
+    // ML job missing "apm" group name
+    return;
+  }
+  const [serviceName] = intersection(apmJobGroups, serviceNames);
+  if (!serviceName) {
+    // APM ML job service was not found
+    return;
+  }
+  const [transactionType] = apmJobGroups.filter(
+    (groupName) => groupName !== serviceName
+  );
+  if (!transactionType) {
+    // APM ML job transaction type was not found.
+    return;
+  }
+  return { jobId: mlJob.job_id, serviceName, transactionType };
+};
+
+export type ServiceAnomalies = PromiseReturnType<typeof getServiceAnomalies>;
+
+export async function getServiceAnomalies(
+  options: IEnvOptions,
+  serviceNames: string[]
+) {
+  const { start, end, ml } = options.setup;
+
+  if (!ml || serviceNames.length === 0) {
+    return [];
+  }
+
+  const { jobs: apmMlJobs } = await ml.anomalyDetectors.jobs('apm');
+  const apmMlJobCategories = apmMlJobs
+    .map((job) => getApmMlJobCategory(job, serviceNames))
+    .filter(
+      (apmJobCategory) => apmJobCategory !== undefined
+    ) as ApmMlJobCategory[];
+  const apmJobIds = apmMlJobs.map((job) => job.job_id);
+  const rangeQuery = { range: rangeFilter(start, end, 'timestamp') };
```

Comment:
I think this is still relevant:
https://github.com/elastic/kibana/pull/65217#discussion_r420484752